x/stakeibc/keeper/unbonding_records.go

```diff
@@ -96,6 +111,74 @@ func (k Keeper) GetTotalUnbondAmountAndRecordsIds(ctx sdk.Context, chainId strin
+) bool {
+	totalNativeAmount := sdkmath.ZeroInt()
+	chainId := hostZoneUnbondingRecord.GetHostZoneId()
+	hostZone, found := k.GetHostZone(ctx, chainId)
+	if !found {
+		k.Logger(ctx).Info(utils.LogWithHostZone(chainId, "No host zone found for id %s, Not Updating Native Token Amounts", chainId))
+		return false
+	}
+	redemptionRate := hostZone.RedemptionRate
+	// loop through user redemption records for the host zone unbonding record, update each one
+	for _, userRedemptionRecordId := range hostZoneUnbondingRecord.GetUserRedemptionRecords() {
+		userRedemptionRecord, found := k.RecordsKeeper.GetUserRedemptionRecord(ctx, userRedemptionRecordId)
+		if !found {
+			k.Logger(ctx).Info(utils.LogWithHostZone(chainId, "No user redemption record found for id %s", userRedemptionRecordId))
+			continue
+		}
+		// update the amount, based on the current redemption rate (or set to zero if desired)
+		nativeAmount := sdkmath.ZeroInt()
+		if !setNativeAmountsToZero {
+			nativeAmount = sdk.NewDecFromInt(userRedemptionRecord.StTokenAmount).Mul(redemptionRate).RoundInt()
+		}
+		userRedemptionRecord.Amount = nativeAmount
+		totalNativeAmount = totalNativeAmount.Add(nativeAmount)
+		k.RecordsKeeper.SetUserRedemptionRecord(ctx, userRedemptionRecord)
+	}
+	// update the host zone unbonding record
+	hostZoneUnbondingRecord.NativeTokenAmount = totalNativeAmount
+	updatedEpochUnbondingRecord, success := k.RecordsKeeper.AddHostZoneToEpochUnbondingRecord(ctx, epochNumber, chainId, hostZoneUnbondingRecord)
+	if !success {
+		// TODO QUESTION: what should we do on failure? do we revert the userRedemptionRecord changes above?
+		k.Logger(ctx).Error(fmt.Sprintf("Failed to set host zone epoch unbonding record: epochNumber %d, chainId %s, hostZoneUnbonding %v", epochNumber, hostZone.ChainId, hostZoneUnbondingRecord))
+		return false
+	}
+	k.RecordsKeeper.SetEpochUnbondingRecord(ctx, *updatedEpochUnbondingRecord)
+
+	return true
+}
+
+// Update Native Amounts in the Host Zone Unbonding Record (and associated User Redemption Records) to 0
+// to reflect that the unbonding has NOT been finalized yet
+func (k Keeper) SetNativeTokensToZeroInUnbondingRecords(ctx sdk.Context, chainId string, epochUnbondingRecordIds []uint64) error {
+	for _, epochUnbondingRecordId := range epochUnbondingRecordIds {
+		k.Logger(ctx).Info(fmt.Sprintf("Updating native token amount on host zone unbondings on EpochUnbondingRecord %d to zero", epochUnbondingRecordId))
+		// fetch the host zone unbonding
+		hostZoneUnbonding, found := k.RecordsKeeper.GetHostZoneUnbondingByChainId(ctx, epochUnbondingRecordId, chainId)
+		if !found {
+			errMsg := fmt.Sprintf("Error fetching host zone unbonding record for epoch: %d, host zone: %s", epochUnbondingRecordId, chainId)
+			k.Logger(ctx).Error(errMsg)
+			return errorsmod.Wrapf(types.ErrHostZoneNotFound, errMsg)
+		}
+		success := k.UpdateNativeTokensForHostZoneUnbondingRecord(ctx, epochUnbondingRecordId, hostZoneUnbonding, true)
+		if !success {
```

Comment:
> 1. Currently, we don't actually need the output of UpdateNativeTokensForHostZoneUnbondingRecord to succeed, it's mainly housekeeping to keep the records neat. If this failed for whatever reason, it would be nice to not halt the chain / break critical functionality

If it succeeded up to and including [this line](https://github.com/Stride-Labs/stride/pull/1049/files/a71b9b47e72f6454a585e88c5cee8a7b24ec6553#diff-5dd4bc8c0f6254ddad2534b66e244b9576d32326538b71c0051ea58eacc6f1e6R355), then `AddHostZoneToEpochUnbondingRecord` failed and we didn't revert, would we end up in a bad state? Afaict the `totalNativeAmount` would be incremented and written to the URR, but the corresponding HZUR would not be updated, so the two would be out of sync.

Reply:
Mentioned in the other thread as well - state does not get reverted in the begin/end blocker even if we return an error. We should wrap this in a temporary context so we can discard state changes if an error is thrown. I believe this is already implemented in an old PR

Reply:
Ah yeah that's a great point Riley! Agreed, I think these will get out-of-sync the way it's currently written

Sam, do you have an example of the temporary context? I can implement it here, I think that's likely the best path forward

Reply:
https://github.com/Stride-Labs/stride/pull/861/files#diff-ba71e9ef6bf1ebcc5243683cbd44405947d8109d7ffb825c7c353d256f7e2179R355-R358

Reply:
same comment as above - this is called outside of the begin blocker (acks, tx to re-open ICAs)